### PR TITLE
Fixed pyright errors inside /memory

### LINKED ISSUE #586: Resolve Pyright errors for Memory

### PR DESCRIPTION
**Pull request by aodhanroche**
Closes #586

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -54,6 +54,7 @@ redis = { version = "^4.6.0", optional = true }
 opensearch-py = { version = "^2.3.1", optional = true }
 pgvector = { version = "^0.2.3", optional = true }
 psycopg2-binary = { version = "^2.9.9", optional = true }
+pyright = "^1.1.349"
```
Comment (collindutter) [inline pyproject.toml:57]: Don't think this should be in this PR.

File: griptape/memory/task/storage/text_artifact_storage.py [hunk 2]
```diff
@@ -18,14 +18,17 @@ class TextArtifactStorage(BaseArtifactStorage):
...
+            raise ValueError("Must be a TextArtifact")
 
     def load_artifacts(self, namespace: str) -> ListArtifact:
         return self.query_engine.load_artifacts(namespace)
 
     def summarize(self, namespace: str) -> TextArtifact:
         return self.summary_engine.summarize_artifacts(self.load_artifacts(namespace))
 
-    def query(self, namespace: str, query: str, metadata: any = None) -> TextArtifact:
+    def query(self, namespace: str, query: str, metadata: Any = None) -> TextArtifact:
```
Comment (collindutter) [inline griptape/memory/task/storage/text_artifact_storage.py:33]: Change `Any` to `dict`

File: griptape/memory/task/storage/text_artifact_storage.py [hunk 3]
```diff
@@ -18,14 +18,17 @@ class TextArtifactStorage(BaseArtifactStorage):
     def can_store(self, artifact: BaseArtifact) -> bool:
         return isinstance(artifact, TextArtifact)
 
-    def store_artifact(self, namespace: str, artifact: TextArtifact) -> None:
-        self.query_engine.upsert_text_artifact(artifact, namespace)
+    def store_artifact(self, namespace: str, artifact: BaseArtifact) -> None:
+        if isinstance(artifact, TextArtifact):
+            self.query_engine.upsert_text_artifact(artifact, namespace)
+        else:
+            raise ValueError("Must be a TextArtifact")
```
Comment (collindutter) [inline griptape/memory/task/storage/text_artifact_storage.py:25]: Update error to `Artifact must be of instance TextArtifact`.

File: griptape/memory/task/storage/blob_artifact_storage.py [hunk 4]
```diff
@@ -10,17 +12,20 @@ class BlobArtifactStorage(BaseArtifactStorage):
...
+            raise ValueError("Must be a BlobArtifact")
 
     def load_artifacts(self, namespace: str) -> ListArtifact:
         return ListArtifact(next((blobs for key, blobs in self.blobs.items() if key == namespace), []))
 
     def summarize(self, namespace: str) -> InfoArtifact:
         return InfoArtifact("can't summarize artifacts")
 
-    def query(self, namespace: str, query: str, metadata: any = None) -> InfoArtifact:
+    def query(self, namespace: str, query: str, metadata: Any = None) -> InfoArtifact:
```
Comment (collindutter) [inline griptape/memory/task/storage/blob_artifact_storage.py:30]: Change `Any` to `dict`

File: griptape/memory/task/storage/blob_artifact_storage.py [hunk 5]
```diff
@@ -10,17 +12,20 @@ class BlobArtifactStorage(BaseArtifactStorage):
...
-            self.blobs[namespace] = []
+    def store_artifact(self, namespace: str, artifact: BaseArtifact) -> None:
+        if isinstance(artifact, BlobArtifact):
+            if namespace not in self.blobs:
+                self.blobs[namespace] = []
 
-        self.blobs[namespace].append(artifact)
+            self.blobs[namespace].append(artifact)
+        else:
+            raise ValueError("Must be a BlobArtifact")
```
Comment (collindutter) [inline griptape/memory/task/storage/blob_artifact_storage.py:22]: Update error to `Artifact must be of instance BlobArtifact`.

File: griptape/memory/task/storage/base_artifact_storage.py [hunk 6]
```diff
@@ -18,9 +20,9 @@ def can_store(self, artifact: BaseArtifact) -> bool:
         ...
 
     @abstractmethod
-    def summarize(self, namespace: str) -> TextArtifact:
+    def summarize(self, namespace: str) -> TextArtifact | InfoArtifact:
         ...
 
     @abstractmethod
-    def query(self, namespace: str, query: str, metadata: any = None) -> TextArtifact:
+    def query(self, namespace: str, query: str, metadata: Any = None) -> TextArtifact | InfoArtifact:
```
Comment (collindutter) [inline griptape/memory/task/storage/base_artifact_storage.py:27]: Update `Any` to `dict`

File: griptape/memory/meta/action_subtask_meta_entry.py [hunk 7]
```diff
@@ -15,6 +18,6 @@ class ActionSubtaskMetaEntry(BaseMetaEntry):
 
     type: str = field(default=BaseMetaEntry.__name__, kw_only=True, metadata={"serializable": False})
 
-    thought: str = field(kw_only=True, metadata={"serializable": True})
+    thought: Optional[str] = field(kw_only=True, metadata={"serializable": True})
```
Comment (collindutter) [inline griptape/memory/meta/action_subtask_meta_entry.py:21]: Don't know if we want this change. What was the pyright error this resolved?

File: poetry.lock
Comment (collindutter) [inline poetry.lock:1]: Don't think this should be in this PR.

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/memory/meta/action_subtask_meta_entry.py (hunk 7 above)
Comment (aodhanroche) [reply griptape/memory/meta/action_subtask_meta_entry.py:21]: This change resolves  /Users/aodhanroche/Documents/GitHub/Griptaping/griptape/griptape/memory/task/task_memory.py:67:56 - error: Argument of type "str | None" cannot be assigned to parameter "thought" of type "str" in function "__init__"
    Type "str | None" cannot be assigned to type "str"
      "None" is incompatible with "str" (reportArgumentType)

File: griptape/memory/task/storage/text_artifact_storage.py (hunk 2 above)
Comment (aodhanroche) [reply griptape/memory/task/storage/text_artifact_storage.py:33]: Elected to go with Any

**Comment by aodhanroche** [comment]
pyright errors are in the unit tests not in /memory

File: tests/unit/memory/tool/test_task_memory.py [hunk 8]
```diff
@@ -57,7 +57,6 @@ def test_process_output(self, memory):
         assert entries[0].answer.startswith(
             'Output of "MockTool.test" was stored in memory with memory_name "MyMemory"'
         )
-        assert entries[0].thought is None
```
Comment (aodhanroche) [inline tests/unit/memory/tool/test_task_memory.py:60]: does this still need to go?

Comment (collindutter) [reply tests/unit/memory/tool/test_task_memory.py:60]: Let's add this back

File: poetry.lock
Comment (collindutter) [inline poetry.lock:1]: Can you fully revert this change?

File: tests/unit/tasks/test_tool_task.py [hunk 9]
```diff
@@ -30,11 +30,14 @@ def test_run_with_memory(self, agent):
...
         agent.add_task(task)
 
+        agent.meta_memory = None
+
         assert task.run().to_text().startswith('Output of "MockTool.test" was stored in memory')
 
     def test_meta_memory(self):
         memory = defaults.text_task_memory("TestMemory")
         subtask = ActionSubtask()
+
```
Comment (collindutter) [inline tests/unit/tasks/test_tool_task.py:40]: Please remove extra newline to reduce PR noise.

File: tests/unit/tasks/test_toolkit_task.py [hunk 10]
```diff
@@ -178,6 +178,7 @@ def test_memory(self, query_engine):
     def test_meta_memory(self):
         memory = defaults.text_task_memory("TestMemory")
         subtask = ActionSubtask()
+
```
Comment (collindutter) [inline tests/unit/tasks/test_toolkit_task.py:181]: Newline

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/memory/meta/action_subtask_meta_entry.py [hunk 11]
```diff
@@ -15,6 +16,6 @@ class ActionSubtaskMetaEntry(BaseMetaEntry):
...
```
Comment (collindutter) [inline griptape/memory/meta/action_subtask_meta_entry.py:19]: Add a `default=None`

File: tests/unit/tasks/test_tool_task.py [hunk 12]
```diff
@@ -30,6 +30,7 @@ def test_run_with_memory(self, agent):
 
         agent.add_task(task)
 
+        agent.meta_memory = None
```
Comment (collindutter) [inline tests/unit/tasks/test_tool_task.py:33]: Is this still necessary with the revert to optional thought?

**Review by collindutter — CHANGES_REQUESTED** [review]

File: tests/unit/tasks/test_tool_task.py (hunk 12 above)
Comment (aodhanroche) [reply tests/unit/tasks/test_tool_task.py]: Don't think is needed anymore, deleted

**Review by collindutter — APPROVED** [review]
